node/src/components/storage.rs

```diff
@@ -143,32 +151,113 @@ impl<B: Value> Default for DeployMetadata<B> {
+            .await
+            .expect("should run");
+
+        // If there is not yet a BlockProposerState stored, we should continue with a default.
+        let mut block_proposer_state = match block_proposer_state {
+            Ok(Some(state)) => state,
+            _ => BlockProposerState::default(),
+        };
+
+        let max_ttl = {
+            let chainspec_store = self.chainspec_store();
+            let chainspec = task::spawn_blocking(move || chainspec_store.get(chainspec_version))
                 .await
-                .expect("should run")
-                .into_iter()
-                .map(|result| {
-                    result.unwrap_or_else(|error| panic!("failed to get deploy: {}", error))
-                })
-                .flatten()
+                .expect("should run blocking");
+            let chainspec = match chainspec {
+                Ok(Some(chainspec)) => chainspec,
+                // If we can't get our hands on a chainspec, then we can't get a max_ttl to compare
+                // blocks and deploys against.
+                _ => return block_proposer_state,
+            };
+            chainspec.genesis.deploy_config.max_ttl
+        };
+
+        // deploys, organized by ProtoBlockHash, which have been finalized
+        let mut finalized = HashMap::new();
+
+        //
+        // - get block by height of latest block
+        'iterate_ancestry: for height in (0..=latest_block_height).rev() {
+            let block = {
+                let block_store = self.block_store();
+                let block_by_height_store = self.block_height_store();
+
+                let ancestor_hash = task::spawn_blocking(move || block_by_height_store.get(height))
+                    .await
+                    .expect("should spawn_blocking");
+                let ancestor_hash = match ancestor_hash {
+                    Ok(Some(hash)) => hash,
+                    _ => break 'iterate_ancestry,
+                };
+
+                let block = task::spawn_blocking(move || block_store.get(smallvec![ancestor_hash]))
+                    .await
+                    .expect("should spawn_blocking")
+                    .pop();
+                match block {
+                    Some(Ok(Some(block))) => block,
+                    _ => break 'iterate_ancestry,
+                }
+            };
+
+            let (block_hash, deploys) = self.load_block_deploys(&block).await;
+
+            {
+                let has_deploys_older_than_max_ttl = deploys
+                    .iter()
+                    .any(|deploy| deploy.header().timestamp() > block.header().timestamp());
+                // if none of the deploys in the block are younger than the block and the block
+                // itself is older than now - max_ttl
+                if !has_deploys_older_than_max_ttl
+                    && block.header().timestamp() > current_instant - max_ttl
```

Comment:
Don't you mean `block.header().timestamp() < current_instant - max_ttl`?
Also, all deploy timestamps being older than the block timestamp should be a _block validity criterion_: Otherwise how do you know that the next block still satisfies it, if you stop iterating here?

Reply:
Nice catch, thanks. If we can rely on the block's _past light cone_ as containing the all the deploys as well, then I'm happy to remove that check. I think I misinterpreted your comment in NDRS-166 as meaning that we should check the block for deploys that might have a timestamp younger than the block.

Reply:
I don't think we currently can rely on that. I'm saying we should add that check: Blocks should be considered illegal if they contain a deploy with a timestamp younger than the block.

Reply:
Should we be panicking here in that case?

Reply:
No, no need to check that here.
We just need to add the check to the validation of incoming blocks, and make sure we only produce valid blocks.

Reply:
So it probably needs to be added to the `BlockValidator` at the point where it has downloaded all the deploys (so it knows their timestamps).